File: webapp/static/physics/lap_sim.js
// JS port of f1sim/lap_sim.py: corner-speed limit -> forward accel pass ->
// backward brake pass, with the friction-ellipse coupling (see f1sim's
// docstring) between lateral and longitudinal grip. Mirrors the Python/
// numba implementation function-for-function.

function interp1(v, xp, fp) {
  const n = xp.length;
  if (v <= xp[0]) return fp[0];
  if (v >= xp[n - 1]) return fp[n - 1];
  let lo = 0, hi = n - 1;
  while (hi - lo > 1) {
    const mid = (lo + hi) >> 1;
    if (xp[mid] <= v) lo = mid; else hi = mid;
  }
  const t = (v - xp[lo]) / (xp[hi] - xp[lo]);
  return fp[lo] + t * (fp[hi] - fp[lo]);
}

function ellipseLongAccel(vi, kappaI, vGrid, aLatGrid, aLongGrid) {
  const aLatMax = interp1(vi, vGrid, aLatGrid);
  const aLongMax = interp1(vi, vGrid, aLongGrid);
  const aLatUsed = vi * vi * Math.abs(kappaI);
  let ratio = aLatMax > 1e-9 ? aLatUsed / aLatMax : 1.0;
  if (ratio > 1.0) ratio = 1.0;
  return aLongMax * Math.sqrt(Math.max(0.0, 1.0 - ratio * ratio));
}

function cornerSpeedLimit(gg, kappa, vMaxGrid = 120.0, iters = 20) {
  const n = kappa.length;
  const out = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    let k = Math.abs(kappa[i]);
    if (k < 1e-6) k = 1e-6;
    let vi = vMaxGrid;
    for (let it = 0; it < iters; it++) {
      const a = interp1(vi, gg.v, gg.a_lat);
      let vNew = Math.sqrt(a / k);
      if (vNew > vMaxGrid) vNew = vMaxGrid;
      vi = 0.5 * vi + 0.5 * vNew;
    }
    out[i] = vi;
  }
  return out;
}

function forwardPass(gg, vCorner, ds, kappa) {
  const n = vCorner.length;
  const v = vCorner.slice();
  for (let i = 0; i < n - 1; i++) {
    const a = ellipseLongAccel(v[i], kappa[i], gg.v, gg.a_lat, gg.a_acc);
    const vLim = Math.sqrt(v[i] * v[i] + 2 * a * ds[i]);
    if (vLim < v[i + 1]) v[i + 1] = vLim;
  }
  return v;
}

function backwardPass(gg, vIn, ds, kappa) {
  const n = vIn.length;
  const v = vIn.slice();
  for (let i = n - 1; i > 0; i--) {
    const a = ellipseLongAccel(v[i], kappa[i], gg.v, gg.a_lat, gg.a_brk);
    const vLim = Math.sqrt(v[i] * v[i] + 2 * a * ds[i - 1]);
    if (vLim < v[i - 1]) v[i - 1] = vLim;
  }
  return v;
}

/** Full lap: corner-limit -> forward -> backward -> lap_time, with lap
 * tiling on closed tracks to remove start/finish transient (mirrors
 * f1sim.lap_sim.simulate_lap). */
function simulateLap(trackR, alpha, gg, closedLaps = 2) {
  const { kappa, path } = trackR.curvatureOfOffset(alpha);
  const ds = trackR.pathLength(alpha);
  const n = trackR.n;

  let v, dsMid;
  if (trackR.closed) {
    const reps = Math.max(closedLaps, 2);
    const vCornerBase = cornerSpeedLimit(gg, kappa);
    const vC = new Float64Array(n * reps);
    const dsRep = new Float64Array(n * reps);
    const kappaRep = new Float64Array(n * reps);
    for (let r = 0; r < reps; r++) {
      vC.set(vCornerBase, r * n);
      dsRep.set(ds, r * n);
      kappaRep.set(kappa, r * n);
    }
    const vF = forwardPass({ v: gg.v, a_lat: gg.a_lat, a_acc: gg.a_acc }, vC, dsRep, kappaRep);
    const vB = backwardPass({ v: gg.v, a_lat: gg.a_lat, a_brk: gg.a_brk }, vF, dsRep, kappaRep);
    const mid = Math.floor(reps / 2);
    v = vB.slice(mid * n, (mid + 1) * n);
    dsMid = ds;
  } else {
    const vCorner = cornerSpeedLimit(gg, kappa);
    const vF = forwardPass(gg, vCorner, ds, kappa);
    v = backwardPass(gg, vF, ds, kappa);
    dsMid = ds;
  }

  const dt = new Float64Array(n);
  let lapTime = 0;
  for (let i = 0; i < n; i++) {
    dt[i] = dsMid[i] / Math.max(v[i], 0.1);
    lapTime += dt[i];
  }

  return { path, kappa, speed: v, ds: dsMid, dt, lap_time: lapTime };
}

File: f1sim/lap_sim.py
"""Quasi-steady-state lap-time march: given a curvature profile (from a
candidate line) and a precomputed GG-diagram, compute the speed profile via
corner-speed limit -> forward accel pass -> backward brake pass, then
integrate lap time.

The per-point recurrences (forward/backward pass, corner-speed fixed point)
are inherently sequential, so they're JIT-compiled with numba: each one
degrades to a tight scalar loop with a hand-rolled binary-search interp,
avoiding per-element numpy/np.interp call overhead. This is the dominant
cost during line optimization since it's re-run on every objective/gradient
evaluation. Falls back to pure numpy if numba isn't installed.
"""
import numpy as np
from . import vehicle_dynamics as vd

try:
    from numba import njit
    _NUMBA = True
except ImportError:
    _NUMBA = False

    def njit(*args, **kwargs):
        if len(args) == 1 and callable(args[0]):
            return args[0]
        return lambda f: f


@njit(cache=True, fastmath=True)
def _interp1(v, xp, fp):
    n = xp.shape[0]
    if v <= xp[0]:
        return fp[0]
    if v >= xp[n - 1]:
        return fp[n - 1]
    lo, hi = 0, n - 1
    while hi - lo > 1:
        mid = (lo + hi) // 2
        if xp[mid] <= v:
            lo = mid
        else:
            hi = mid
    x0, x1 = xp[lo], xp[hi]
    f0, f1 = fp[lo], fp[hi]
    t = (v - x0) / (x1 - x0)
    return f0 + t * (f1 - f0)


@njit(cache=True, fastmath=True)
def _corner_speed_limit_core(v_grid, a_lat_grid, kappa, v_max_grid, iters):
    n = kappa.shape[0]
    out = np.empty(n)
    for i in range(n):
        k = abs(kappa[i])
        if k < 1e-6:
            k = 1e-6
        vi = v_max_grid
        for _ in range(iters):
            a = _interp1(vi, v_grid, a_lat_grid)
            v_new = np.sqrt(a / k)
            if v_new > v_max_grid:
                v_new = v_max_grid
            vi = 0.5 * vi + 0.5 * v_new
        out[i] = vi
    return out


@njit(cache=True, fastmath=True)
def _forward_pass_core(v_grid, a_acc_grid, v_corner, ds):
    n = v_corner.shape[0]
    v = v_corner.copy()
    for i in range(n - 1):
        a = _interp1(v[i], v_grid, a_acc_grid)
        v_lim = np.sqrt(v[i] ** 2 + 2 * a * ds[i])
        if v_lim < v[i + 1]:
            v[i + 1] = v_lim
    return v


@njit(cache=True, fastmath=True)
def _backward_pass_core(v_grid, a_brk_grid, v_in, ds):
    n = v_in.shape[0]
    v = v_in.copy()
    for i in range(n - 1, 0, -1):
        a = _interp1(v[i], v_grid, a_brk_grid)
        v_lim = np.sqrt(v[i] ** 2 + 2 * a * ds[i - 1])
        if v_lim < v[i - 1]:
            v[i - 1] = v_lim
    return v


def corner_speed_limit(gg, kappa, v_max_grid=120.0, iters=20):
    """v such that a_lat_max(v) == v^2 * |kappa|, solved per-point via
    fixed-point iteration (a_lat_max is a mild function of v)."""
    kappa = np.ascontiguousarray(kappa, dtype=np.float64)
    return _corner_speed_limit_core(gg["v"], gg["a_lat"], kappa, float(v_max_grid), int(iters))


def forward_pass(gg, v_corner, ds):
    v_corner = np.ascontiguousarray(v_corner, dtype=np.float64)
    ds = np.ascontiguousarray(ds, dtype=np.float64)
    return _forward_pass_core(gg["v"], gg["a_acc"], v_corner, ds)


def backward_pass(gg, v, ds):
    v = np.ascontiguousarray(v, dtype=np.float64)
    ds = np.ascontiguousarray(ds, dtype=np.float64)
    return _backward_pass_core(gg["v"], gg["a_brk"], v, ds)


def warmup(gg):
    """Trigger numba JIT compilation once, up front, so the cost doesn't
    land inside the first optimizer iteration."""
    if not _NUMBA:
        return
    dummy_k = np.array([1e-3, 1e-2])
    dummy_ds = np.array([1.0, 1.0])
    v = corner_speed_limit(gg, dummy_k)
    v = forward_pass(gg, v, dummy_ds)
    backward_pass(gg, v, dummy_ds)


def simulate_lap(track_r, alpha, car, gg, closed_laps=2):
    """Returns dict with speed profile, lap time, path points, curvature.
    For closed tracks, runs the forward/backward pass over `closed_laps`
    concatenated laps and takes the middle lap's time to remove start/finish
    transient effects (a proper cyclic solve would iterate until v wraps
    consistently; this concatenation approximates that cheaply)."""
    kappa, path = track_r.curvature_of_offset(alpha)
    ds = track_r.path_length(alpha)

    v_corner = corner_speed_limit(gg, kappa)

    if track_r.closed:
        reps = max(closed_laps, 2)
        v_c = np.tile(v_corner, reps)
        ds_rep = np.tile(ds, reps)
        v_f = forward_pass(gg, v_c, ds_rep)
        v_b = backward_pass(gg, v_f, ds_rep)
        n = track_r.n
        mid = reps // 2
        v = v_b[mid * n:(mid + 1) * n]
        ds_mid = ds
    else:
        v_f = forward_pass(gg, v_corner, ds)
        v = backward_pass(gg, v_f, ds)
        ds_mid = ds

    dt = ds_mid / np.maximum(v, 0.1)
    lap_time = float(np.sum(dt))

    return {
        "path": path,
        "kappa": kappa,
        "speed": v,
        "ds": ds_mid,
        "dt": dt,
        "lap_time": lap_time,
    }
